internal/ansible/ansible_manager.go

```diff
@@ -0,0 +1,390 @@
+	errorCode = UndefError
+	errorDetails = event.Stdout
+	if strings.Contains(fmt.Sprintf("%v", errorDetails), "The command was not found or was not executable: ansible-playbook") {
+		errorCode = NotInstalled
+	}
+	// TODO: enumerate more failure types
+	return errorCode, errorDetails
+}
+
+// execPlaybook executes the ansible playbook.
+// It sends ansible playbook results when the playbook execution has been completed on playbookResults channel.
+// When the execution terminates, execPlaybook signals on executionCompleted channel if there was an error.
+func execPlaybook(
+	executionCompleted chan error,
+	playbookResults chan<- *ansibleResults.AnsiblePlaybookJSONResults,
+	playbookCmd playbook.AnsiblePlaybookCmd,
+	timeout time.Duration,
+	messageID string,
+	buffOut *bytes.Buffer,
+	mappingRepository mapping.MappingRepository,
+	modTime time.Time) {
+
+	defer close(executionCompleted)
+	defer close(playbookResults)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Check if the playbook has been already executed on the device.
+	alreadyExecuted := mappingRepository.Exists(modTime)
+
+	if alreadyExecuted {
+		err := fmt.Errorf("playbook of messageID %s is already in execution", messageID)
+		executionCompleted <- err
+		return
+	}
+
+	fileContent, err := os.ReadFile(playbookCmd.Playbooks[0])
+	if err != nil {
+		executionCompleted <- err
+		return
+	}
+
+	err = mappingRepository.Add(fileContent, modTime)
+	if err != nil {
+		executionCompleted <- err
+		return
+	}
+	errRun := playbookCmd.Run(ctx)
+
+	if errRun != nil {
+		log.Warnf("playbook executed with errors. Results: %s, messageID: %s, Error: %v", buffOut.String(), messageID, errRun)
+	}
+	results, err := ansibleResults.JSONParse(buffOut.Bytes())
+	if err != nil {
+		log.Errorf("error while parsing json string %s. MessageID: %s, Error: %v\n", buffOut.String(), messageID, err)
+		// Signal that the playbook execution completed with error
+		executionCompleted <- err
+		// No more work to be done, return
+		return
+	}
+
+	playbookResults <- results
+
+	err = mappingRepository.Remove(fileContent)
+	if err != nil {
+		log.Errorf("cannot remove pending playbook %s. MessageID: %s, Error: %v\n", string(fileContent), messageID, err)
+	}
+
+	// Signal that the playbook execution completed
+	executionCompleted <- errRun
+}
+
+// execPlaybookSync executes the ansible playbook synchronously.
+// if error is not nil, then an error occured during the playbook execution (e.g. host unreachable), but this does't mean
+// that there are no results available. In other words, a successful call returns *ansibleResults.AnsiblePlaybookJSONResults not nil.
+func (a *Manager) execPlaybookSync(
+	playbook *playbook.AnsiblePlaybookCmd,
+	timeout time.Duration,
+	buffOut *bytes.Buffer,
+	mappingRepository mapping.MappingRepository) (*ansibleResults.AnsiblePlaybookJSONResults, error) {
```

Comment:
`mappingRepository` is unused here, but I think the playbook should be removed from it. when the execution ends.